### PR TITLE
[WIP] Put database operations in separate package

### LINKED ISSUE #2: Put database operations in separate package
IMHO it is good to put database logic in separate package. This way We can support more databases in future. Also it makes code more maintainable.

### PR DESCRIPTION
**Pull request by vbmade2000**
Fixes #2

### CODE REVIEW

File: database/boltdb.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
+package database
+
+import (
+	"bytes"
+	"errors"
+	bolt "github.com/coreos/bbolt"
```
Comment (sbinet) [inline database/boltdb.go:6]: please use `goimports` to format your code.

https://golang.org/x/tools/cmd/goimports

File: database/boltdb.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
+package database
```
Comment (sbinet) [inline database/boltdb.go:1]: we are missing a copyright header here.

File: database/boltdb.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	"errors"
+	bolt "github.com/coreos/bbolt"
+	"sort"
+)
+
+// BoltDataStore contains methods that deal with BoltDB
+type BoltDataStore struct {
+	Db *bolt.DB
+}
+type byteSlice [][]byte
```
Comment (sbinet) [inline database/boltdb.go:14]: I would move this closer to the methods defined on that type :)

File: database/common/common.go [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package common
```
Comment (sbinet) [inline database/common/common.go:1]: we are missing a copyright header here as well.

File: database/common/common.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+
+type DB interface {
+   Subscribers(list string) ([]string, error)
+   Subscribe(user, list string) error
+   Usubscribe(user, list string) error
+   Lists() ([]string, error)
+   Users() ([]string, error)
+}
```
Comment (sbinet) [inline database/common/common.go:9]: we are missing a `\n` there.

File: server.go [hunk 6]
```diff
@@ -31,10 +32,11 @@ var (
...
 type Server struct {
-	cfg Config
-	db  *bolt.DB
-	sck net.Listener
-	msg chan *Message
+	cfg       Config
+	db        *bolt.DB
+	sck       net.Listener
+	msg       chan *Message
+	boltStore *database.BoltDataStore
```
Comment (sbinet) [inline server.go:39]: using the explicit type here kind of defeats the purpose of introducing an interface :)

we should drop the `db *bolt.DB` field and replace `boltStore` with:

```go
type Server struct {
    // ...
    db database.Store
}
```

assuming the interface `Store` is defined in the `database` package:

```go
package database

type Store interface {
     Subscribers(list string) ([]string, error)
￼     Subscribe(user, list string) error
     Usubscribe(user, list string) error
     Lists() ([]string, error)
     Users() ([]string, error)
}
```

then, the config should be modified to specify the database name/type to use.
name to be used later on with your `database.New()` or, rather: `database.Open("boltdb", cfg.Database)`

File: database/boltdb.go (hunk 1 above)
Comment (vbmade2000) [reply database/boltdb.go:6]: Sure. Actually entire formatting is yet to be done.

File: database/boltdb.go (hunk 2 above)
Comment (vbmade2000) [reply database/boltdb.go:1]: I'll add that.

File: database/boltdb.go (hunk 3 above)
Comment (vbmade2000) [reply database/boltdb.go:14]: Sure.

File: database/common/common.go (hunk 4 above)
Comment (vbmade2000) [reply database/common/common.go:1]: I'll add that.

File: database/common/common.go (hunk 5 above)
Comment (vbmade2000) [reply database/common/common.go:9]: Yes. I'll check that too.

File: server.go (hunk 6 above)
Comment (vbmade2000) [reply server.go:39]: Agreed with that. Conrete type is to be replaced by interface but BoltDBStore didn't implement all the methods of interface so it was remaining. I'll do that now.

**Comment by vbmade2000** [comment]
I think We should also add isSubscribed() to Store interface and move it to boltdb.go. What do you think ?

**Comment by sbinet** [comment]
this can be a helper function implemented in terms of the `Store` interface:

```go
func isSubscribed(db Store, user, list string) bool {
    users, err := db.Subscribers(list)
    if err != nil { return false }
    for _, u := range users {
        if u == user { return true }
    }
    return false
}
```

no need to add it to the interface.

**Comment by vbmade2000** [comment]
I modified as following.
```
func (srv *Server) isSubscribed(db common.Store, user, list string) bool {
	var o bool
	usr := []byte(user)
	key := []byte(list)
	err := db.Update(func(tx *bolt.Tx) error {
		b := tx.Bucket(subBucket)
		v := b.Get(key)
		for _, v := range bytes.Split(v, []byte(",")) {
			if bytes.Equal(v, usr) {
				o = true
				return nil
			}
		}
		return nil
	})
	if err != nil {
		return false
	}
	return o
}
```

It raises `db.Update undefined.(type common.Store has no method or field Db)`

I can add GetDB() method to Store interface but not sure is it a proper way.

**Comment by sbinet** [comment]
I must say I don't understand how you get that error.

also: as I said in https://github.com/sbinet-alt63/strew/pull/7#issuecomment-372648049 `isSubscribed` can be implemented in terms of `Store.Subscribers()`.
no need to reach into `boltdb.Update` and its buckets.

**Comment by vbmade2000** [comment]
It already has db.Update() call. I just added `db common.Store` in function arguments.

**Comment by sbinet** [comment]
closed in favour of #10

**Pull request closed without merging**